drivers/adc/adc_stm32.c

```diff
@@ -798,93 +798,77 @@ static void adc_stm32_setup_speed(const struct device *dev, uint8_t id,
 
 static void adc_stm32_setup_channels(const struct device *dev, uint8_t channel_id)
 {
-#if defined(CONFIG_SOC_SERIES_STM32G4X) || \
-	defined(CONFIG_SOC_SERIES_STM32F3X) || \
-	defined(CONFIG_SOC_SERIES_STM32F3X) || \
-	defined(CONFIG_SOC_SERIES_STM32F7X) || \
-	defined(CONFIG_SOC_SERIES_STM32H7X) || \
-	defined(CONFIG_SOC_SERIES_STM32L4X) || \
-	defined(CONFIG_SOC_SERIES_STM32L5X) || \
-	defined(CONFIG_SOC_SERIES_STM32MP1X)
 	const struct adc_stm32_cfg *config = dev->config;
 	ADC_TypeDef *adc = config->base;
-#endif
 
-#if defined(CONFIG_SOC_SERIES_STM32G4X)
-	/* In the G4 model, once can access internal temperature on ADCs 1 and 5. */
-	if ((adc == ADC1 &&
-		__LL_ADC_CHANNEL_TO_DECIMAL_NB(ADC_CHANNEL_TEMPSENSOR_ADC1) == channel_id) ||
-		(adc == ADC5 &&
-		__LL_ADC_CHANNEL_TO_DECIMAL_NB(ADC_CHANNEL_TEMPSENSOR_ADC5) == channel_id)) {
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(adc1), okay) && \
+	DT_PROP(DT_NODELABEL(adc1), has_temp_channel)
+	if (adc == ADC1 &&
+		__LL_ADC_CHANNEL_TO_DECIMAL_NB(ADC_CHANNEL_TEMPSENSOR) == channel_id) {
```

Comment:
From this point, you should be able to do a "mere" (modulo the series api adaptation):
```
	if (config->has_temp_channel &&
		__LL_ADC_CHANNEL_TO_DECIMAL_NB(ADC_CHANNEL_TEMPSENSOR) == channel_id) {
```
This requires adding fields to device config struct, but likely worse the gain  considering this function.